server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -648,8 +650,27 @@ private IndexShard getShard(ShardSearchRequest request) {
         return indicesService.indexServiceSafe(request.shardId().getIndex()).getShard(request.shardId().id());
     }
 
-    private static <T> void runAsync(Executor executor, CheckedSupplier<T, Exception> executable, ActionListener<T> listener) {
-        executor.execute(ActionRunnable.supply(listener, executable));
+    private static <T extends RefCounted> void runAsync(
+        Executor executor,
+        CheckedSupplier<T, Exception> executable,
+        ActionListener<T> listener
+    ) {
+        executor.execute(ActionRunnable.wrap(listener, new CheckedConsumer<>() {
+            @Override
+            public void accept(ActionListener<T> l) throws Exception {
+                var res = executable.get();
+                try {
```

Comment:
Also, following this pattern (instantiate -> try -> finally -> decRef) as closely as possible and not implicitly requiring listeners to ever count down a ref makes the code a lot easier to follow and extend (as can be seen in the test simplifications ... mod the futures :P).